[runtime-audit-engine] New module (#3477)
deckhouse/deckhouse: ee/modules/650-runtime-audit-engine/docs/README.md

---

> title: "The runtime-audit-engine module"

Author: There are different approaches to module naming. I tried to be in line with the other security module - admission policy engine.

---

## Overview

The module implements a runtime threats detection engine. 
It can collect Linux kernel system calls and Kubernetes API audit events, enrich them with metadata from Kubernetes Pods and generate security audit events according to conditional rules.

This module:
* Detects threats at runtime by observing the behavior of your applications and containers.
* Helps to detect CVEs exploits and cryptocurrency mining attacks. 
* Improves Kubernetes security by detecting:
  * Shells running in containers or Pods in Kubernetes.
  * Containers running in privileged mode or attempting to mount sensitive paths, such as `/proc`, on the host.
  * Unauthorized attempts to read confidential files such as `/etc/shadow`.

## Architecture

The module is based on the [Falco](https://falco.org/) system. 
Deckhouse deploys Falco agents (which run as a DaemonSet) on every node. The agents then start consuming kernel / kube audit events.

![Falco DaemonSet](../../images/650-runtime-audit-engine/falco_daemonset.png)

> > NOTE: We recommend deploying Falco as a systemd unit for maximum security.

Author: There is a mistake after grammar and style fixes. I will rewrite this block.

> > NOTE: We recommend deploying Falco as a systemd unit for maximum security.

Author: Fixed

> However, a Kubernetes cluster with the autoscaling feature enabled makes it hard to operate. 

> > Additional security mechanisms such as multitenancy and admission policy control provide the required level of security to mitigate attacks on the Falco DaemonSet.

Author: It makes sens to me. I will add a note.

> > Additional security mechanisms such as multitenancy and admission policy control provide the required level of security to mitigate attacks on the Falco DaemonSet.

Author: Added

A single Pod consists of five containers:
![Falco Pod](../../images/650-runtime-audit-engine/falco_pod.png)

1. `falco-driver-loader` — init container that compiles eBPF program and stores it into empty dir to provide access to it for Falco.
2. `falco` — collects events, enriches them with metadata and stores them.

> 3. `rules-loader` — collects custom resources (`FalcoAuditRules`) from Kubernetes and store them in a shared directory (empty dir).

Reviewer: ```suggestion
3. `rules-loader` — collects custom resources ([FalcoAuditRules](cr.html#falcoauditrules)) from Kubernetes and stores them in a shared directory (empty dir).
```

4. `falcosidekick`: only exports events as metrics to be able to alert on them.
5. `kube-rbac-proxy`: protects falcosidekick metrics endpoint.

## Audit Rules

> An ability to collect events on its own means nothing, because the amount of data that can be exported from a Linux kernel is too much to analyze by a human brain.

Reviewer 2: ```suggestion
An ability to collect events on its own means nothing, because the amount of data that can be exported from a Linux kernel is too big to analyze by a human.
```

The amount is countable.

There is quite a bit of obvious mistakes overall, and I believe you can easily find them yourself. I decided not to pay much attention to them. I guess it can be corrected in a dedicated iteration.

Rules address this problem by collecting events according to certain pattens that can help in detecting suspicious activities.

The main part of a rule is a conditional expression (which uses the [conditions syntax](https://falco.org/docs/rules/conditions/)).

### Embedded rules

There are several built-in rules that cannot be disabled.  
These rules are aimed at detecting Deckhouse security problems as well as security problems affecting the `runtime-audit-engine` module.

- `/etc/falco/falco_rules.yaml` — syscall rules;
- `/etc/falco/k8s_audit_rules.yaml` — Kubernetes audit rules.

### Custom audit rules

Users can use a `FalcoAuditRules` CRD to add custom security audit rules. 

> Each Falco agent Pod has a sidecar container running [shell-operator](https://github.com/flant/shell-operator).

Author: I think it's not worth it because it makes documentation harder to maintain. IMO, this description as it is is exemplary.

> Each Falco agent Pod has a sidecar container running [shell-operator](https://github.com/flant/shell-operator).

Reviewer 2: To me it's like speaking about one thing in two different ways. It does not make it easier for a readers. I don't insist anyway.

This sidecar reads rules from the custom resources and saves them in the Pod's `/etc/falco/rules.d/` directory.
Falco automatically reloads the configuration when a new rule becomes available.

![Falco shell-operator](../../images/650-runtime-audit-engine/falco_shop.png)

Such a schema allows the IaC approach to be used to maintain Falco rules.

## Requirements

### OS

The module uses the eBPF Falco driver to ingest syscall data. It is better suited for environments where loading a kernel module is prohibited or unsupported, such as GKE, EKS, and other managed Kubernetes solutions.
However, there are known limitations to the eBPF driver:

> * The eBPF probe may not work for every system.

Author: It is always about the kernel and its capabilities. In addition to outdated kernels, there may be custom kernels and systems for which capabilities to deploy eBPF may be banned.

> * At least Linux kernel version 4.14 is required, but the Falco project suggests an LTS kernel of 4.14/4.19 or above.

Reviewer 3: Should we add init container which checks kernel version  like in node-local-dns module ?

Author: Honestly, I didn't check how falco works in every environment. This requirement is copied from the Falco doc.

Reviewer 3: If falco uses bpf minimal kernel version is 4.16. This is not a blocking point, may be add this to TODO ?

Author: Added to TODO list

### CPU / Memory

> Falco agents are deployed on every node. Resources consumption of each Pod depends on the number of rules / ingested events.

Author: Fixed.

> Falco agents are deployed on every node. Resources consumption of each Pod depends on the number of rules / ingested events.

Reviewer 2: I left another comment here!

> Falco agents are running on every node. Therefore, the resource consumption of each Pod depends on the number of rules or ingested events.

Reviewer 2: > depends on the number of rules or ingested events

Again, is it really OR, or, maybe, it is AND?

Author: Or, yes.

## Kubernetes Audit Webhook

[Webhook audit mode](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/#webhook-backend) should be configured to collect audit events of kube-apiserver. 
If the control-plane-manager module is enabled, settings will be automatically applied when the runtime-audit-engine module is enabled.

For Kubernetes cluster with control-plane out of Deckhouse control, it is possible to manually configure the webhook.
1. Create a webhook kubeconfig file with address https://127.0.0.1:9765/k8s-audit and CA (ca.crt) from the `d8-runtime-audit-engine/runtime-audit-engine-webhook-tls` secret.
    
    Example:
    ```yaml
    apiVersion: v1
    kind: Config
    clusters:
    - name: webhook
      cluster:
        certificate-authority-data: BASE64_CA
        server: "https://127.0.0.1:9765/k8s-audit"
    users:
    - name: webhook
    contexts:
    - context:
       cluster: webhook
       user: webhook
      name: webhook
    current-context: webhook
    ```
2. Add the `--audit-webhook-config-file` flag to kube-apiserver manifest that points to previously created file.

> > NOTE: Do not forget to configure audit policy, because, be default, Deckhouse only collect audit events from system namespaces.

Reviewer 3: ```suggestion
> NOTE: Do not forget to configure audit policy, because, by default, Deckhouse only collect audit events from system namespaces.
```

> > NOTE: Remember to configure the audit policy, because Deckhouse only collects audit events from the system namespaces by default.

Author: This NOTE is not about Falco audit events, but about audit events of Kubernetes API collected by Falco. 
1. I clarified the note.
2. Added TODO to manage audit configuration with Falco automatically.

---
title: "The runtime-audit-engine module"
---

## Overview

The module implements a runtime threats detection engine. 
It can collect Linux kernel system calls and Kubernetes API audit events, enrich them with metadata from Kubernetes Pods and generate security audit events according to conditional rules.

This module:
* Detects threats at runtime by observing the behavior of your applications and containers.
* Helps to detect CVEs exploits and cryptocurrency mining attacks. 
* Improves Kubernetes security by detecting:
  * Shells running in containers or Pods in Kubernetes.
  * Containers running in privileged mode or attempting to mount sensitive paths, such as `/proc`, on the host.
  * Unauthorized attempts to read confidential files such as `/etc/shadow`.

## Architecture

The module is based on the [Falco](https://falco.org/) system. 
Deckhouse deploys Falco agents (which run as a DaemonSet) on every node. The agents then start consuming kernel / kube audit events.

![Falco DaemonSet](../../images/650-runtime-audit-engine/falco_daemonset.svg)
<!--- Source: https://docs.google.com/drawings/d/1NZ91z8NXNiuS50ybcMoMsZI3SbQASZXJGLANdaNNm_U --->

> > NOTE: Falco developers recommend deploying Falco as a systemd unit for maximum security.

Reviewer: ```suggestion
> Falco developers recommend deploying Falco as a systemd unit for maximum security.
```

<!--- Source: https://docs.google.com/drawings/d/13MFYtiwH4Y66SfEPZIcS7S2wAY6vnKcoaztxsmX1hug --->

Such a schema allows the IaC approach to be used to maintain Falco rules.

## Requirements

### OS

The module uses the eBPF Falco driver to ingest syscall data. It is better suited for environments where loading a kernel module is prohibited or unsupported, such as GKE, EKS, and other managed Kubernetes solutions.
However, there are known limitations to the eBPF driver:
* The eBPF probe may not work for every system.
* At least a Linux kernel version 4.14 is required, although the Falco project suggests an LTS kernel version 4.14/4.19 or higher.

### CPU / Memory

Falco agents are running on every node. Therefore, the resource consumption of each Pod depends on the number of rules or ingested events.

## Kubernetes Audit Webhook

[Webhook audit mode](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/#webhook-backend) should be configured to collect audit events of `kube-apiserver`. 
If the [control-plane-manager](../040-control-plane-manager/) module is enabled, settings will be automatically applied when the `runtime-audit-engine` module is enabled.

> You can manually configure the webhook for Kubernetes clusters with a control plane that is not controlled by Deckhouse.

Reviewer: ```suggestion
You can manually configure the webhook for Kubernetes clusters with a control plane that is not controlled by Deckhouse:
```
